### ISSUE TITLE
Write acknowledgement not called if value unchanged

### ISSUE THREAD

**Issue by n-scope** [deepstreamIO/deepstream.io#539]
Thanks for your great work at deepstream.
I am not sure if it's a bug or an enhancement, but for the sake of simplicity, I think the acknowledgement callback should **always** be called.

- Using the official docker image with rethinkdb and deepstream 2.1.2, the callback is never called **if the record value is left unchanged**:

``` javascript
dsClient.record.getRecord("testAcknowledgement");
let record = dsClient.record.getRecord("testAcknowledgement");

record.set({
        message: "hello"
    }, e => {
        if (e) {
            console.log(e);
        }
        else {
            console.log("Acknowledgement received for hello");
        }
    }
);

record.set({
        message: "hello"
    }, e => {
        if (e) {
            console.log(e);
        }
        else {
            console.log("Acknowledgement received for second hello");
        }
    }
);
// prints only 'Acknowledgement received for hello'
```
- As a side-effect, how are we supposed to discard a record using write aknowledgment ?
``` javascript
record.set({
        message: "hello"
    }, e => {
        // Not sure it will be called (depends on the record value) 
        record.discard();
    }
);

record.set({
        message: "hello"
    }, e => {}
);
// We get an UNSOLICITED_MESSAGE error, because we discard too soon
record.discard();
```

Thanks a lot !

**Comment by WolframHempel** [comment]
@n-scope thanks :-) - the problem is that in order to improve efficiency the client checks for changes before sending out messages - for unchanged values no interaction with the server happens (and therefor no write). I understand your usecase, but I am worried that we are watering down the semantics of "write acknowledgements" if there hasn't been an actual write to acknowledge

**Comment by ronag** [comment]
@WolframHempel: Shouldn't at least the client call the callback if it is not sending it to the server?

**Comment by WolframHempel** [comment]
that's what I mean - that means acknowledging a write that hasn't happened

**Comment by WolframHempel** [comment]
it's also worth adding that messages are queued internally and write ack listeners are optional. The following would work with an implicit write ack

```
record.set('some', 'value');
record.discard();
```

**Comment by n-scope** [comment]
thanks for your answer :) 

I understand your point of view, acknowledging an unexistant write seems awkward. 
But since the callback is already called when a write error occurs (which is very useful!), I think it is already more than a write acknowledement.

In my opinion, the efficiency optimization performed by deepstream should be transparent, all we need to know is that the backend is in sync and that everything went fine.

Is it mandatory to discard, or is there any internal timeout which cleans the records up ? 
If so, is it possible to both use the write acknowledement feature and discard properly the record ?

**Comment by yasserf** [comment]
> Is it mandatory to discard records, or is there any internal timeout which cleans the records up ?

Yes it is. Alot of clients just subscribe to updates and never actually set data. Because of that you'll need to ensure you discard your records during your applications life cycle events

**Comment by yasserf** [comment]
This has been fixed in 2.1.3! Woohoo
